maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java

```diff
@@ -0,0 +1,231 @@
+
+        this.eventQueue = new ArrayBlockingQueue<>(queueSize);
+        Thread updater = new Thread(this::feedConsumer);
+        updater.setDaemon(true);
+        updater.start();
+    }
+
+    public TransferListener getDelegate() {
+        return delegate;
+    }
+
+    private void feedConsumer() {
+        final ArrayList<Exchange> batch = new ArrayList<>(batchMaxSize);
+        try {
+            while (true) {
+                batch.clear();
+                if (eventQueue.drainTo(batch, BATCH_MAX_SIZE) == 0) {
+                    batch.add(eventQueue.take());
+                }
+                demux(batch);
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void demux(List<Exchange> exchanges) {
+        for (Exchange exchange : exchanges) {
+            exchange.process(transferEvent -> {
+                TransferEvent.EventType type = transferEvent.getType();
+                try {
+                    switch (type) {
+                        case INITIATED:
+                            delegate.transferInitiated(transferEvent);
+                            break;
+                        case STARTED:
+                            delegate.transferStarted(transferEvent);
+                            break;
+                        case PROGRESSED:
+                            delegate.transferProgressed(transferEvent);
+                            break;
+                        case CORRUPTED:
+                            delegate.transferCorrupted(transferEvent);
+                            break;
+                        case SUCCEEDED:
+                            delegate.transferSucceeded(transferEvent);
+                            break;
+                        case FAILED:
+                            delegate.transferFailed(transferEvent);
+                            break;
+                        default:
+                            LOGGER.warn("Invalid TransferEvent.EventType={}; ignoring it", type);
+                    }
+                } catch (TransferCancelledException e) {
+                    ongoing.put(transferEvent.getResource().getFile(), Boolean.FALSE);
+                }
+            });
+        }
+    }
+
+    private void put(TransferEvent event, boolean last) {
+        try {
+            Exchange exchange;
+            if (blockOnLastEvent && last) {
+                exchange = new BlockingExchange(event);
+            } else {
+                exchange = new Exchange(event);
+            }
+            eventQueue.put(exchange);
+            exchange.waitForProcessed();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final ConcurrentHashMap<File, Boolean> ongoing = new ConcurrentHashMap<>();
+
+    @Override
+    public void transferInitiated(TransferEvent event) {
+        ongoing.putIfAbsent(event.getResource().getFile(), Boolean.TRUE);
```

Comment:
@cstamas , `TransferResource` supports `file` being `null` ([ref](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/transfer/TransferResource.java#L166)). Althought the field is deprecated, the replacement is also nullable ([ref](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/transfer/TransferResource.java#L178)). However, this listener implementation will break if the resource has no `file` set. Was it a conscious decision?

Reply:
No, this is a probably oversight... care to provide a PR for this? And thanks!

Reply:
Yeah, I can do that. I'll look into it in https://issues.apache.org/jira/browse/MNG-8150